### PR TITLE
Change WriteEvent calls to WriteEventCore

### PR DESCRIPTION
**Pull request by ToddGrun**
[WriteEvent ](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventsource.writeevent?view=net-8.0#system-diagnostics-tracing-eventsource-writeevent(system-int32-system-object())) is quite inefficient in that it takes in a params array of objects, thus allocating both the array and boxing the int/long parameters passed in. Instead, WriteEventCore should be called with an appropriately populated EventSource.EventData.

Found while investigating OOP allocations in a speedometer test.

Old Memory Allocations from speedometer
image

New Memory Allocations from speedometer
image

### CODE REVIEW

**Comment by ToddGrun** [comment]
Created as a draft PR as I'd like to get a test insertion with this change to look at whether this has the desired impact

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs [hunk 1]
```diff
@@ -38,12 +37,63 @@ public static class Tasks
...
+                {
+                    EventSource.EventData* data = stackalloc EventSource.EventData[4];
+                    data[0].DataPointer = (IntPtr)generatorNameBytes;
+                    data[0].Size = ((generatorName.Length + 1) * 2);
+                    data[1].DataPointer = (IntPtr)assemblyPathBytes;
+                    data[1].Size = ((assemblyPath.Length + 1) * 2);
+                    data[2].DataPointer = (IntPtr)(&elapsedTicks);
+                    data[2].Size = 8;
+                    data[3].DataPointer = (IntPtr)idBytes;
+                    data[3].Size = ((id.Length + 1) * 2);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CodeAnalysisEventSource.cs:56]:
```suggestion
                    data[0].Size = ((generatorName.Length + 1) * sizeof(char));
                    data[1].DataPointer = (IntPtr)assemblyPathBytes;
                    data[1].Size = ((assemblyPath.Length + 1) * sizeof(char));
                    data[2].DataPointer = (IntPtr)(&elapsedTicks);
                    data[2].Size = sizeof(long);
                    data[3].DataPointer = (IntPtr)idBytes;
                    data[3].Size = ((id.Length + 1) * sizeof(char));
```

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs [hunk 2]
```diff
@@ -38,12 +37,63 @@ public static class Tasks
         internal void StartSingleGeneratorRunTime(string generatorName, string assemblyPath, string id) => WriteEvent(3, generatorName, assemblyPath, id);
 
         [Event(4, Message = "Generator {0} ran for {2} ticks", Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Stop, Task = Tasks.SingleGeneratorRunTime)]
-        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id) => WriteEvent(4, generatorName, assemblyPath, elapsedTicks, id);
+        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id)
+        {
+            unsafe
+            {
+                fixed (char* generatorNameBytes = generatorName)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CodeAnalysisEventSource.cs:44]: 💭 I think I prefer this form as a way to ensure buffer lengths always match up:

```cs
fixed (char* generatorNameBytes = generatorName)
fixed (char* assemblyPathBytes = assemblyPath)
fixed (char* idBytes = id)
{
    Span<EventData> data = stackalloc EventData[]
    {
        new() { DataPointer = (IntPtr)generatorNameBytes, Size = (generatorName.Length + 1) * sizeof(char) },
        new() { DataPointer = (IntPtr)assemblyPathBytes, Size = (assemblyPath.Length + 1) * sizeof(char) },
        new() { DataPointer = (IntPtr)(&elapsedTicks), Size = sizeof(long) },
        new() { DataPointer = (IntPtr)idBytes, Size = (id.Length + 1) * sizeof(char) },
    };

    fixed (EventData* dataPtr = data)
    {
        WriteEventCore(eventId: 4, data.Length, dataPtr);
    }
}
```

Comment (sharwell) [reply src/Compilers/Core/Portable/CodeAnalysisEventSource.cs:44]: Could go one step further:

```cs
fixed (char* generatorNameBytes = generatorName)
fixed (char* assemblyPathBytes = assemblyPath)
fixed (char* idBytes = id)
{
    Span<EventData> data = stackalloc EventData[]
    {
        eventDataForString(generatorName, generatorNameBytes),
        eventDataForString(assemblyPath, assemblyPathBytes),
        eventDataForInt64(&elapsedTicks),
        eventDataForString(id, idBytes),
    };

    fixed (EventSource.EventData* dataPtr = data)
    {
        WriteEventCore(eventId: 4, data.Length, dataPtr);
    }
}

static EventData eventDataForString(string value, char* ptr)
{
    fixed (char* ptr2 = value)
    {
        if (ptr2 != ptr)
            throw new ArgumentException("Pinned value must match string.");
    }

    return new EventData()
    {
        DataPointer = (IntPtr)ptr,
        Size = (value.Length + 1) * sizeof(char),
    };
}

static EventData eventDataForInt64(long* ptr)
{
    return new EventData()
    {
        DataPointer = (IntPtr)ptr,
        Size = sizeof(long),
    };
}
```

Comment (ToddGrun) [reply src/Compilers/Core/Portable/CodeAnalysisEventSource.cs]: Can you take a look at the 2nd commit. Followed your suggestions fairly closely as it's a lot better than the framework's EventSource code that I originally mimic'ed.

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs [hunk 3]
```diff
@@ -65,35 +62,56 @@ internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPa
...
+            {
+                return new EventData()
+                {
+                    DataPointer = (IntPtr)ptr,
+                    Size = (value.Length + 1) * sizeof(char),
+                };
+            }
+        }
+
+        private static EventData GetEventDataForInt(int value)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CodeAnalysisEventSource.cs]: 1. Use the runtime type for the method name
2. Caller must pass in the pointer

```suggestion
        private static EventData GetEventDataForInt32(int* value)
```

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs [hunk 4]
```diff
@@ -65,35 +62,56 @@ internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPa
...
+        {
+            return new EventData()
+            {
+                DataPointer = (IntPtr)value,
+                Size = sizeof(int),
+            };
+        }
+
+        private static EventData GetEventDataForLong(long value)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CodeAnalysisEventSource.cs]:
```suggestion
        private static EventData GetEventDataForInt64(long* value)
```

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs [hunk 5]
```diff
@@ -65,35 +62,56 @@ internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPa
...
             }
         }
+
+        private static unsafe EventData GetEventDataForString(string value)
+        {
+            fixed (char* ptr = value)
+            {
+                return new EventData()
+                {
+                    DataPointer = (IntPtr)ptr,
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CodeAnalysisEventSource.cs]: ❗ This is going to unpin `value` before returning. This must be refactored so the caller pins the value.

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs (hunk 4 above)
Comment (ToddGrun) [reply src/Compilers/Core/Portable/CodeAnalysisEventSource.cs]: Was addressed in commit 3

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs [hunk 6]
```diff
@@ -38,12 +37,95 @@ public static class Tasks
...
+        }
 
         [Event(5, Message = "Generator '{0}' failed with exception: {1}", Level = EventLevel.Error)]
         internal void GeneratorException(string generatorName, string exception) => WriteEvent(5, generatorName, exception);
 
         [Event(6, Message = "Node {0} transformed", Keywords = Keywords.Correctness, Level = EventLevel.Verbose, Task = Tasks.BuildStateTable)]
-        internal void NodeTransform(int nodeHashCode, string name, string tableType, int previousTable, string previousTableContent, int newTable, string newTableContent, int input1, int input2) => WriteEvent(6, nodeHashCode, name, tableType, previousTable, previousTableContent, newTable, newTableContent, input1, input2);
+        internal void NodeTransform(int nodeHashCode, string name, string tableType, int previousTable, string previousTableContent, int newTable, string newTableContent, int input1, int input2)
+        {
+            unsafe
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CodeAnalysisEventSource.cs:70]: 💡 Two more suggestions:

1. Move `unsafe` modifier to method
2. Change this line to `if (IsEnabled())`

**Review by sharwell — APPROVED** [review]

**Review by sharwell — APPROVED** [review]

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler -- ptal

File: eng/config/BannedSymbols.txt [hunk 7]
```diff
@@ -69,4 +69,5 @@ M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetOperationsAsync(System.Thread
 M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetPreviewOperationsAsync(System.Threading.CancellationToken); Use overload that takes a solution
 M:Microsoft.CodeAnalysis.CodeActions.CodeAction.ComputeOperationsAsync(System.Threading.CancellationToken); Use overload that takes progress
 M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionAsync(CSystem.Threading.CancellationToken); Use overload that takes progress
```
Comment (sharwell) [inline eng/config/BannedSymbols.txt:71]:
```suggestion
M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionAsync(System.Threading.CancellationToken); Use overload that takes progress
```

File: eng/config/BannedSymbols.txt [hunk 8]
```diff
@@ -69,4 +69,5 @@ M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetOperationsAsync(System.Thread
...
-M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedDocumentAsync(CancellationToken); Use overload that takes progress
\ No newline at end of file
+M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedDocumentAsync(CancellationToken); Use overload that takes progress
```
Comment (sharwell) [inline eng/config/BannedSymbols.txt:72]:
```suggestion
M:Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedDocumentAsync(System.Threading.CancellationToken); Use overload that takes progress
```

File: eng/config/BannedSymbols.txt (hunk 7 above)
Comment (ToddGrun) [reply eng/config/BannedSymbols.txt:71]: Not making this change (or the other one right below) as there are usages in CodeAction.cs that are outside the scope of this PR.

**Comment by ToddGrun** [comment]
ping @dotnet/roslyn-compiler -- ptal

**Comment by ToddGrun** [comment]
@333fred , @jaredpar could you please take a look? The change is pretty trivial. Thanks!

**Comment by jaredpar** [comment]
> The change is pretty trivial. 

Changes with that much `unsafe` are never trivial

**Comment by ToddGrun** [comment]
> Changes with that much `unsafe` are never trivial

Yeah, that was just my way to sucker you into looking at it :)

**Comment by ToddGrun** [comment]
> Changes with that much `unsafe` are never trivial

This change is very closely based on what the non-params WriteEvent* methods do in System.Diagnostics.Tracing.EventSource. From the overload taking in params in documentation: 

This is a varargs helper for writing an event. It does create an array and box all the arguments so it is relatively inefficient and should only be used for relatively rare events

**Review by jaredpar — APPROVED** [review]

**Comment by jaredpar** [comment]
@chsienki PTAL

**Review by chsienki — APPROVED** [review]

**Pull request merged**